File: Y2Q1/SDB/Project/Code/Scripts/pbfe2/pbfe_lib/DistanceJoin.scala
package main

import org.apache.spark.{SparkConf, SparkContext}
import org.apache.spark.sql.{SQLContext, Row}
import org.apache.spark.sql.Point

object DistanceJoin {
  case class PointItem(id: Int, x: Double, y: Double)
  case class Pair(id1: Int, id2: Int, x1: Double, y1: Double, x2: Double, y2: Double)

  def main(args: Array[String]) : Unit = {
    val sparkConf = new SparkConf().setAppName("DistanceJoin").setMaster("local[*]")
    val sc = new SparkContext(sparkConf)
    val sqlContext = new SQLContext(sc)

    import sqlContext.implicits._

    val filename = "P10K.csv"
    val epsilon = 100

    val p1 = sc.textFile(filename).map(_.split(",")).map(p => PointItem(p(0).trim.toInt, p(1).trim.toDouble, p(2).trim.toDouble)).toDF()
    val p2 = p1.toDF("id2", "x2", "y2")

    val pairs = p1.distanceJoin(p2, Point(p1("x"), p1("y")), Point(p2("x2"), p2("y2")), epsilon)
    val disks = pairs.rdd
      .filter( (x:Row) => x.getInt(0) > x.getInt(3) )
      .map( (x: Row) => calculateDisks(x) )
    println(disks.count())

    sc.stop()
  }
}

File: Y2Q1/SDB/Project/Code/Scripts/misc/DistanceJoin.py
from pyspark import SparkConf, SparkContext
from pyspark.sql import SQLContext
from pyspark.sql import Row

conf = (SparkConf()\
	.setMaster("local")\
	.setAppName("My app")\
	.set("spark.executor.memory", "1g"))
sc = SparkContext(conf = conf)
sqlContext = SQLContext(sc)

epsilon = 10
points = sc.textFile("P10K.csv")\
	.map(lambda line: line.split(","))\
	.map(lambda p: Row(id=p[0], lat=float(p[1]), lng=float(p[2])))\
	.toDF()

npoints = points.count()
npoints

points.registerTempTable("p1")
points.registerTempTable("p2")

sql = """
    SELECT 
        * 
    FROM 
        p1 
    DISTANCE JOIN 
        p2 
    ON 
        POINT(p2.lng, p2.lat) IN CIRCLERANGE(POINT(p1.lng, p1.lat), {0}) 
    WHERE 
        p2.id < p1.id""".format(epsilon)

pairs = sqlContext.sql(sql)
print pairs.count()


